server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1302,78 +1404,126 @@ public Entitlement adjustEntitlementQuantity(Consumer consumer,
         // Because there are several paths to this one place where entitlements
         // are granted, we cannot be positive the caller obtained a lock on the
-        // pool
-        // when it was read. As such we're going to reload it with a lock
+        // pool when it was read. As such we're going to reload it with a lock
         // before starting this process.
-        log.info("Locking pool: {}", pool.getId());
-        pool = poolCurator.lockAndLoad(pool);
 
-        if (quantity > 0) {
+        log.debug("Locking pools: {}", poolQuantityMap.keySet());
+
+        List<Pool> pools = poolCurator.lockAndLoad(poolQuantityMap.keySet());
+
+        if (log.isDebugEnabled()) {
+            for (Pool pool : pools) {
+                log.debug("Locked pool: {} consumed: {}", pool, pool.getConsumed());
+            }
+        }
+
+        Map<String, PoolQuantity> poolQuantities = new HashMap<String, PoolQuantity>();
+        boolean quantityFound = false;
+        for (Pool pool : pools) {
+            Integer quantity = poolQuantityMap.get(pool.getId());
+            if (quantity > 0) {
+                quantityFound = true;
+            }
+            poolQuantities.put(pool.getId(), new PoolQuantity(pool, quantity));
+            poolQuantityMap.remove(pool.getId());
+        }
+
+        if (!poolQuantityMap.isEmpty()) {
+            throw new IllegalArgumentException(i18n.tr("Subscription pool(s) {0} do not exist.",
+                    poolQuantityMap.keySet()));
+        }
+        if (quantityFound) {
             log.info("Running pre-entitlement rules.");
             // XXX preEntitlement is run twice for new entitlement creation
-            ValidationResult result = enforcer.preEntitlement(consumer, pool, quantity, caller);
-
-            if (!result.isSuccessful()) {
-                log.warn("Entitlement not granted: {}", result.getErrors().toString());
-                throw new EntitlementRefusedException(result);
+            Map<String, ValidationResult> results = enforcer.preEntitlement(consumer,
+                    poolQuantities.values(), caller);
+
+            for (Entry<String, ValidationResult> entry : results.entrySet()) {
+                ValidationResult result = entry.getValue();
+                if (!result.isSuccessful()) {
+                    log.warn("Entitlement not granted: {} for pool: {}",
+                            result.getErrors().toString(), entry.getKey());
+                    throw new EntitlementRefusedException(results);
+                }
             }
         }
+
         EntitlementHandler handler = null;
-        if (entitlement == null) {
+        if (entitlements == null) {
             handler = new NewHandler();
         }
+        else if (entitlements.keySet().size() != poolQuantities.keySet().size() ||
+                !entitlements.keySet().containsAll(poolQuantities.keySet()) ||
+                !poolQuantities.keySet().containsAll(entitlements.keySet())) {
```

Comment:
Since keySets are of type `Set`, could we just use `!poolQuantities.keySet().equals(entitlements.keySet())`?